lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs

```diff
@@ -413,6 +419,187 @@ spec = do
             property $ \(a :: Int) (b :: Int) ->
                 compare (InclusiveBound a) (InclusiveBound b) === compare a b
 
+    describe "Epoch arithmetic: predecessors and successors" $ do
+
+        let succN n = applyN n (epochSucc =<<)
+        let predN n = applyN n (epochPred =<<)
+
+        it "epochPred minBound == Nothing" $
+            epochPred minBound === Nothing
+
+        it "epochSucc maxBound == Nothing" $
+            epochSucc maxBound === Nothing
+
+        it "(applyN n epochSucc) . (applyN n epochPred) == id" $
+            withMaxSuccess 1000 $ property $
+                \(Small epochWord) (Small n) ->
+                    let epoch = EpochNo epochWord
+                        withinBounds = minBound + n <= unEpochNo epoch
+                        expectedResult =
+                            if withinBounds then Just epoch else Nothing
+                    in
+                    checkCoverage $
+                    cover 10      withinBounds  "within bounds" $
+                    cover 10 (not withinBounds) "out of bounds" $
+                    expectedResult === succN n (predN n $ Just epoch)
+
+        it "(applyN n epochPred) . (applyN n epochSucc) == id" $
+            withMaxSuccess 1000 $ property $
+                \(Small epochWord) (Small n) ->
+                    let epoch = EpochNo $ maxBound - epochWord
+                        withinBounds = maxBound - n >= unEpochNo epoch
+                        expectedResult =
+                            if withinBounds then Just epoch else Nothing
+                    in
+                    checkCoverage $
+                    cover 10      withinBounds  "within bounds" $
+                    cover 10 (not withinBounds) "out of bounds" $
+                    expectedResult === predN n (succN n $ Just epoch)
```

Comment:
**[question]**

Wouldn't be better to add properties in those two boundary cases where we apply n times `epochPred` and n times `epochPred` but randomly picked and also require certain coverage of out-of-bounds cases?

Reply:
> **[suggestion]**
> 
> Wouldn't be better to add properties in those two boundary cases where we apply n times `epochPred` and n times `epochSucc` but randomly picked (rather than one special case n times `epochPred` and later `times`epochSucc`) and also require certain coverage of out-of-bounds cases?

I'm not sure I understand what you mean here. Would you be able to give an example?